File: packages/sites/src/protocol.ts
import {
  canonical,
  exactShape,
  type Identity,
  type PublicIdentity,
} from "../../core/src/protocol";
import type { CertificateCrypto } from "../../core/src/certificate-types";

export const SITE_REVISION_LIMITS = Object.freeze({
  certificateBytes: 4096,
  predecessors: 16,
  headers: 1024,
  sequence: Number.MAX_SAFE_INTEGER,
  payloadBytes: 3_500_000,
});
export interface SiteRevisionBody {
  domain: "relayloom/site-snapshot/1";
  owner: PublicIdentity;
  name: string;
  number: number;
  previous: string[];
  documentHash: string;
}
export interface SiteRevision {
  body: SiteRevisionBody;
  id: string;
  signature: string;
}
export interface SiteAddress {
  ownerId: string;
  name: string;
}
function requireThat(value: unknown, reason: string): asserts value {
  if (!value) throw new Error(reason);
}
function address(value: unknown): asserts value is string {
  requireThat(
    typeof value === "string" && /^[a-f0-9]{64}$/.test(value),
    "Identificador de site inválido",
  );
}
function siteName(value: unknown): asserts value is string {
  requireThat(
    typeof value === "string" && /^[a-z][a-z0-9-]{0,39}$/.test(value),
    "Nome estável de site inválido",
  );
}
function list(value: unknown, maximum: number): asserts value is unknown[] {
  requireThat(
    Array.isArray(value) &&
      Object.getPrototypeOf(value) === Array.prototype &&
      value.length <= maximum,
    "Lista de revisões fora dos limites",
  );
  const descriptors = Object.getOwnPropertyDescriptors(value);
  requireThat(
    Reflect.ownKeys(value).length === value.length + 1 &&
      Array.from(
        { length: value.length },
        (_, i) => descriptors[String(i)],
      ).every((d) => d && "value" in d),
    "Lista de revisões inválida",
  );
}
export function siteAddress(ownerId: string, name: string): string {
  address(ownerId);
  siteName(name);
  return "relayloom:site:" + ownerId + "/" + name;
}
export function parseSiteAddress(value: unknown): SiteAddress {
  requireThat(
    typeof value === "string" && value.length <= 120,
    "Endereço de site inválido",
  );
  const match = /^relayloom:site:([a-f0-9]{64})\/([a-z][a-z0-9-]{0,39})$/.exec(
    value,
  );
  requireThat(match, "Endereço de site inválido");
  return { ownerId: match[1], name: match[2] };
}

/** Every revision authorises a complete snapshot, not a patch requiring lost
 * ancestors. Previous IDs retain verifiable history when available. The owner
 * explicitly supersedes lower-numbered snapshots; no wall clock grants authority.
 * Storage/adoption, operation idempotence and stale-draft CAS are separate layers. */
export function createSiteRevisionProtocol(crypto: CertificateCrypto) {
  const encoder = new TextEncoder();
  function bounded(value: unknown, maximum: number) {
    const text = canonical(value);
    requireThat(
      encoder.encode(text).length <= maximum,
      "Revisão de site demasiado grande",
    );
    return text;
  }
  function body(value: unknown): asserts value is SiteRevisionBody {
    requireThat(
      exactShape(value, [
        "domain",
        "owner",
        "name",
        "number",
        "previous",
        "documentHash",
      ]),
      "Campos da revisão inválidos",
    );
    const candidate = value as SiteRevisionBody;
    requireThat(
      candidate.domain === "relayloom/site-snapshot/1",
      "Domínio de assinatura de site inválido",
    );
    requireThat(
      crypto.validateIdentity(candidate.owner),
      "Proprietário de site inválido",
    );
    siteName(candidate.name);
    requireThat(
      Number.isSafeInteger(candidate.number) &&
        candidate.number >= 1 &&
        candidate.number <= SITE_REVISION_LIMITS.sequence,
      "Número de revisão inválido",
    );
    address(candidate.documentHash);
    list(candidate.previous, SITE_REVISION_LIMITS.predecessors);
    for (const id of candidate.previous) address(id);
    requireThat(
      candidate.previous.every(
        (id, i) => i === 0 || candidate.previous[i - 1] < id,
      ),
      "Antecessores de site não canónicos",
    );
    requireThat(
      candidate.number !== 1 || candidate.previous.length === 0,
      "A primeira revisão não tem antecessores",
    );
    bounded(candidate, SITE_REVISION_LIMITS.certificateBytes);
  }
  function signature(value: unknown): Uint8Array {
    requireThat(
      typeof value === "string" && value.length === 88,
      "Assinatura de revisão inválida",
    );
    const bytes = Uint8Array.from(atob(value), (character) =>
      character.charCodeAt(0),
    );
    requireThat(
      bytes.length === 64 && btoa(String.fromCharCode(...bytes)) === value,
      "Assinatura de revisão não canónica",
    );
    return bytes;
  }
  function verifyRevision(value: unknown): SiteRevision {
    requireThat(
      exactShape(value, ["body", "id", "signature"]),
      "Certificado de site inválido",
    );
    const revision = value as SiteRevision;
    body(revision.body);
    address(revision.id);
    const signed = bounded(
      revision.body,
      SITE_REVISION_LIMITS.certificateBytes,
    );
    requireThat(
      revision.id === crypto.hash(signed),
      "Hash da revisão não coincide",
    );
    requireThat(
      crypto.verify(revision.body.owner, signed, signature(revision.signature)),
      "Assinatura de site inválida",
    );
    requireThat(
      !revision.body.previous.includes(revision.id),
      "Revisão de site circular",
    );
    bounded(revision, SITE_REVISION_LIMITS.certificateBytes);
    return revision;
  }
  function createRevision(
    identity: Identity,
    name: string,
    number: number,
    previous: string[],
    documentHash: string,
  ): SiteRevision {
    list(previous, SITE_REVISION_LIMITS.predecessors);
    const value: SiteRevisionBody = {
      domain: "relayloom/site-snapshot/1",
      owner: identity.public,
      name,
      number,
      previous: [...previous],
      documentHash,
    };
    body(value);
    value.owner = structuredClone(value.owner);
    const text = bounded(value, SITE_REVISION_LIMITS.certificateBytes);
    const result: SiteRevision = {
      body: value,
      id: crypto.hash(text),
      signature: crypto.sign(identity, text),
    };
    verifyRevision(result);
    return result;
  }
  function createSuccessor(
    identity: Identity,
    name: string,
    parents: SiteRevision[],
    documentHash: string,
  ) {
    siteName(name);
    list(parents, SITE_REVISION_LIMITS.predecessors);
    for (const parent of parents) {
      verifyRevision(parent);
      requireThat(
        parent.body.owner.id === identity.public.id &&
          parent.body.name === name,
        "Antecessor de outro proprietário ou site",
      );
    }
    const number =
      Math.max(0, ...parents.map((parent) => parent.body.number)) + 1;
    return createRevision(
      identity,
      name,
      number,
      [...new Set(parents.map((parent) => parent.id))].sort(),
      documentHash,
    );
  }
  /** Hash only a fully validated complete authored payload. Resource/schema
   * validation must happen before calling this helper or rendering the payload. */
  function documentHash(payload: unknown) {
    return crypto.hash(
      bounded(
        { domain: "relayloom/site-document/1", payload },
        SITE_REVISION_LIMITS.payloadBytes,
      ),
    );
  }
  function verifySnapshot(
    value: unknown,
    payload: unknown,
    ownerId: string,
    name: string,
  ): SiteRevision {
    address(ownerId);
    siteName(name);
    const revision = verifyRevision(value);
    requireThat(
      revision.body.owner.id === ownerId && revision.body.name === name,
      "A revisão pertence a outro site",
    );
    requireThat(
      revision.body.documentHash === documentHash(payload),
      "O documento não corresponde à revisão assinada",
    );
    return revision;
  }
  function historyLink(child: SiteRevision, parent: SiteRevision) {
    requireThat(
      child.body.owner.id === parent.body.owner.id &&
        child.body.name === parent.body.name &&
        parent.body.number < child.body.number &&
        child.body.previous.includes(parent.id),
      "Ligação de histórico inválida",
    );
  }
  function verifyHistoryLink(child: SiteRevision, parent: SiteRevision) {
    verifyRevision(child);
    verifyRevision(parent);
    historyLink(child, parent);
  }
  /** Inputs include persisted verified headers whose payload may have been
   * evicted. A missing newer payload must never silently reveal an older version.
   * A head classification alone is not permission to render: verifySnapshot binds
   * the actual payload, and the application must also verify its transport bundle. */
  function classifyRevisions(ownerId: string, name: string, values: unknown) {
    address(ownerId);
    siteName(name);
    list(values, SITE_REVISION_LIMITS.headers);
    const revisions = new Map<string, SiteRevision>();
    for (const value of values) {
      const revision = verifyRevision(value);
      requireThat(
        revision.body.owner.id === ownerId && revision.body.name === name,
        "Histórico de outro site",
      );
      revisions.set(revision.id, revision);
    }
    for (const revision of revisions.values())
      for (const id of revision.body.previous) {
        const parent = revisions.get(id);
        if (parent) historyLink(revision, parent);
      }
    if (!revisions.size)
      return {
        status: "empty" as const,
        number: 0,
        heads: [] as SiteRevision[],
        missingHistory: [] as string[],
      };
    const number = Math.max(
      ...[...revisions.values()].map((r) => r.body.number),
    );
    const heads = [...revisions.values()]
      .filter((r) => r.body.number === number)
      .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
    const missing = new Set<string>(),
      seen = new Set<string>(),
      queue = [...heads];
    while (queue.length) {
      const revision = queue.pop()!;
      if (seen.has(revision.id)) continue;
      seen.add(revision.id);
      for (const id of revision.body.previous) {
        const parent = revisions.get(id);
        if (parent) queue.push(parent);
        else missing.add(id);
      }
    }
    return {
      status: heads.length === 1 ? ("head" as const) : ("conflict" as const),
      number,
      heads,
      missingHistory: [...missing].sort(),
    };
  }
  return {
    createRevision,
    createSuccessor,
    verifyRevision,
    documentHash,
    verifySnapshot,
    verifyHistoryLink,
    classifyRevisions,
  };
}

File: native/sites/protocol.go
// Package sites verifies owner-signed complete snapshots. Persisted adoption,
// stale-draft CAS, transport and rendering are separate application boundaries.
package sites

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"math"
	"regexp"
	"sort"
	"strings"

	"github.com/JohnnyPBelo/relayloom/native/core"
)

const CertificateBytes = 4096
const PayloadBytes = 3500000
const Predecessors = 16
const Headers = 1024
const MaxSequence int64 = 9007199254740991
const domain = "relayloom/site-snapshot/1"

var namePattern = regexp.MustCompile(`^[a-z][a-z0-9-]{0,39}$`)

type Body struct {
	Domain       string              `json:"domain"`
	Owner        core.PublicIdentity `json:"owner"`
	Name         string              `json:"name"`
	Number       int64               `json:"number"`
	Previous     []string            `json:"previous"`
	DocumentHash string              `json:"documentHash"`
}
type Revision struct {
	Body      Body   `json:"body"`
	ID        string `json:"id"`
	Signature string `json:"signature"`
}
type Selection struct {
	Status         string     `json:"status"`
	Number         int64      `json:"number"`
	Heads          []Revision `json:"heads"`
	MissingHistory []string   `json:"missingHistory"`
}

func Address(ownerID, name string) (string, error) {
	if !core.ValidAddress(ownerID) || !namePattern.MatchString(name) {
		return "", errors.New("endereço de site inválido")
	}
	return "relayloom:site:" + ownerID + "/" + name, nil
}
func ParseAddress(value string) (string, string, error) {
	if len(value) > 120 || !strings.HasPrefix(value, "relayloom:site:") {
		return "", "", errors.New("endereço de site inválido")
	}
	parts := strings.Split(strings.TrimPrefix(value, "relayloom:site:"), "/")
	if len(parts) != 2 {
		return "", "", errors.New("endereço de site inválido")
	}
	if _, err := Address(parts[0], parts[1]); err != nil {
		return "", "", err
	}
	return parts[0], parts[1], nil
}
func bounded(value any, maximum int) ([]byte, error) {
	data, err := core.Canonical(value)
	if err != nil {
		return nil, err
	}
	if len(data) > maximum {
		return nil, errors.New("revisão de site demasiado grande")
	}
	return data, nil
}
func validBody(body Body) error {
	if body.Domain != domain || !namePattern.MatchString(body.Name) || body.Number < 1 || body.Number > MaxSequence || !core.ValidAddress(body.DocumentHash) || body.Previous == nil || len(body.Previous) > Predecessors {
		return errors.New("revisão de site inválida")
	}
	if err := core.ValidateIdentity(body.Owner); err != nil {
		return err
	}
	for i, id := range body.Previous {
		if !core.ValidAddress(id) || (i > 0 && body.Previous[i-1] >= id) {
			return errors.New("antecessores de site não canónicos")
		}
	}
	if body.Number == 1 && len(body.Previous) != 0 {
		return errors.New("a primeira revisão não tem antecessores")
	}
	_, err := bounded(body, CertificateBytes)
	return err
}
func VerifyRevision(revision Revision) error {
	if err := validBody(revision.Body); err != nil {
		return err
	}
	data, err := bounded(revision.Body, CertificateBytes)
	if err != nil {
		return err
	}
	if !core.ValidAddress(revision.ID) || revision.ID != core.Hash(data) || len(revision.Signature) != 88 {
		return errors.New("certificado de site inválido")
	}
	signature, err := base64.StdEncoding.Strict().DecodeString(revision.Signature)
	if err != nil || len(signature) != 64 || base64.StdEncoding.EncodeToString(signature) != revision.Signature {
		return errors.New("assinatura de revisão não canónica")
	}
	if err = core.VerifyCertificateData(revision.Body.Owner, data, signature); err != nil {
		return err
	}
	for _, id := range revision.Body.Previous {
		if id == revision.ID {
			return errors.New("revisão de site circular")
		}
	}
	_, err = bounded(revision, CertificateBytes)
	return err
}
func CreateRevision(identity core.Identity, name string, number int64, previous []string, hash string) (Revision, error) {
	var empty Revision
	if len(previous) > Predecessors {
		return empty, errors.New("antecessores fora dos limites")
	}
	body := Body{domain, identity.Public, name, number, append([]string{}, previous...), hash}
	if err := validBody(body); err != nil {
		return empty, err
	}
	data, err := bounded(body, CertificateBytes)
	if err != nil {
		return empty, err
	}
	signature, err := core.SignCertificateData(identity, data)
	if err != nil {
		return empty, err
	}
	result := Revision{Body: body, ID: core.Hash(data), Signature: base64.StdEncoding.EncodeToString(signature)}
	if err = VerifyRevision(result); err != nil {
		return empty, err
	}
	return result, nil
}
func CreateSuccessor(identity core.Identity, name string, parents []Revision, hash string) (Revision, error) {
	if len(parents) > Predecessors {
		return Revision{}, errors.New("antecessores fora dos limites")
	}
	var highest int64
	ids := map[string]bool{}
	for _, parent := range parents {
		if err := VerifyRevision(parent); err != nil {
			return Revision{}, err
		}
		if parent.Body.Owner.ID != identity.Public.ID || parent.Body.Name != name {
			return Revision{}, errors.New("antecessor de outro proprietário ou site")
		}
		if parent.Body.Number > highest {
			highest = parent.Body.Number
		}
		ids[parent.ID] = true
	}
	if highest >= MaxSequence {
		return Revision{}, errors.New("número de revisão esgotado")
	}
	previous := make([]string, 0, len(ids))
	for id := range ids {
		previous = append(previous, id)
	}
	sort.Strings(previous)
	return CreateRevision(identity, name, highest+1, previous, hash)
}
func DocumentHash(payload any) (string, error) {
	data, err := bounded(map[string]any{"domain": "relayloom/site-document/1", "payload": payload}, PayloadBytes)
	if err != nil {
		return "", err
	}
	return core.Hash(data), nil
}
func VerifySnapshot(revision Revision, payload any, ownerID, name string) error {
	if _, err := Address(ownerID, name); err != nil {
		return err
	}
	if err := VerifyRevision(revision); err != nil {
		return err
	}
	if revision.Body.Owner.ID != ownerID || revision.Body.Name != name {
		return errors.New("a revisão pertence a outro site")
	}
	hash, err := DocumentHash(payload)
	if err != nil {
		return err
	}
	if hash != revision.Body.DocumentHash {
		return errors.New("o documento não corresponde à revisão assinada")
	}
	return nil
}
func historyLink(child, parent Revision) error {
	if child.Body.Owner.ID != parent.Body.Owner.ID || child.Body.Name != parent.Body.Name || parent.Body.Number >= child.Body.Number {
		return errors.New("ligação de histórico inválida")
	}
	for _, id := range child.Body.Previous {
		if id == parent.ID {
			return nil
		}
	}
	return errors.New("antecessor não referenciado")
}
func VerifyHistoryLink(child, parent Revision) error {
	if err := VerifyRevision(child); err != nil {
		return err
	}
	if err := VerifyRevision(parent); err != nil {
		return err
	}
	return historyLink(child, parent)
}
func ClassifyRevisions(ownerID, name string, values []Revision) (Selection, error) {
	result := Selection{Status: "empty", Heads: []Revision{}, MissingHistory: []string{}}
	if _, err := Address(ownerID, name); err != nil {
		return result, err
	}
	if len(values) > Headers {
		return result, errors.New("histórico de site demasiado grande")
	}
	byID := map[string]Revision{}
	for _, value := range values {
		if err := VerifyRevision(value); err != nil {
			return result, err
		}
		if value.Body.Owner.ID != ownerID || value.Body.Name != name {
			return result, errors.New("histórico de outro site")
		}
		byID[value.ID] = value
		if value.Body.Number > result.Number {
			result.Number = value.Body.Number
		}
	}
	for _, value := range byID {
		for _, id := range value.Body.Previous {
			if parent, ok := byID[id]; ok {
				if err := historyLink(value, parent); err != nil {
					return result, err
				}
			}
		}
	}
	if len(byID) == 0 {
		return result, nil
	}
	for _, value := range byID {
		if value.Body.Number == result.Number {
			result.Heads = append(result.Heads, value)
		}
	}
	sort.Slice(result.Heads, func(i, j int) bool { return result.Heads[i].ID < result.Heads[j].ID })
	result.Status = "head"
	if len(result.Heads) > 1 {
		result.Status = "conflict"
	}
	queue := append([]Revision{}, result.Heads...)
	seen, missing := map[string]bool{}, map[string]bool{}
	for len(queue) > 0 {
		value := queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		if seen[value.ID] {
			continue
		}
		seen[value.ID] = true
		for _, id := range value.Body.Previous {
			if parent, ok := byID[id]; ok {
				queue = append(queue, parent)
			} else {
				missing[id] = true
			}
		}
	}
	for id := range missing {
		result.MissingHistory = append(result.MissingHistory, id)
	}
	sort.Strings(result.MissingHistory)
	return result, nil
}

func object(value any, keys ...string) (map[string]any, error) {
	m, ok := value.(map[string]any)
	if !ok || len(m) != len(keys) {
		return nil, errors.New("campos da revisão inválidos")
	}
	for _, key := range keys {
		if _, ok = m[key]; !ok {
			return nil, errors.New("campo da revisão ausente")
		}
	}
	return m, nil
}

// DecodeRevision rejects extra/missing/wrongly typed fields before converting to
// a typed struct; ordinary json.Unmarshal would silently discard unknown fields.
func DecodeRevision(data []byte) (Revision, error) {
	var empty Revision
	value, err := core.DecodeJSON(data, CertificateBytes)
	if err != nil {
		return empty, err
	}
	outer, err := object(value, "body", "id", "signature")
	if err != nil {
		return empty, err
	}
	b, err := object(outer["body"], "domain", "owner", "name", "number", "previous", "documentHash")
	if err != nil {
		return empty, err
	}
	texts := make([]string, 5)
	for i, value := range []any{outer["id"], outer["signature"], b["domain"], b["name"], b["documentHash"]} {
		var ok bool
		texts[i], ok = value.(string)
		if !ok {
			return empty, errors.New("texto de revisão inválido")
		}
	}
	number, ok := b["number"].(json.Number)
	if !ok {
		return empty, errors.New("número de revisão inválido")
	}
	n, err := number.Float64()
	if err != nil || math.IsNaN(n) || math.IsInf(n, 0) || n != math.Trunc(n) || n < 1 || n > float64(MaxSequence) {
		return empty, errors.New("número de revisão inválido")
	}
	previous, ok := b["previous"].([]any)
	if !ok || len(previous) > Predecessors {
		return empty, errors.New("antecessores inválidos")
	}
	ids := make([]string, len(previous))
	for i, value := range previous {
		ids[i], ok = value.(string)
		if !ok {
			return empty, errors.New("antecessor inválido")
		}
	}
	card, err := bounded(b["owner"], CertificateBytes)
	if err != nil {
		return empty, err
	}
	owner, err := core.DecodePublicIdentity(card)
	if err != nil {
		return empty, err
	}
	revision := Revision{Body: Body{texts[2], owner, texts[3], int64(n), ids, texts[4]}, ID: texts[0], Signature: texts[1]}
	if err = VerifyRevision(revision); err != nil {
		return empty, err
	}
	return revision, nil
}
